File: iosApp/iosApp/Modules/CurrencyPicker/CurrencyPickerPresenter.swift
// Created by web3d4v on 06/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

enum CurrencyPickerPresenterEvent {
    case search(searchTerm: String)
    case selectNetwork(CurrencyPickerViewModel.Network)
    case selectCurrency(CurrencyPickerViewModel.Currency)
    case addCustomCurrency
    case viewWillDismiss
    case done
    case dismiss
}

protocol CurrencyPickerPresenter {
    func present()
    func handle(_ event: CurrencyPickerPresenterEvent)
}

final class DefaultCurrencyPickerPresenter {
    private weak var view: CurrencyPickerView?
    private let wireframe: CurrencyPickerWireframe
    private let interactor: CurrencyPickerInteractor
    private let context: CurrencyPickerWireframeContext
    
    private var searchTerm: String = ""
    private var selectedNetwork: Network!
    private var networks: [Network] = []
    private var selectedCurrencies: [Currency] = []
    private var selectedCurrenciesFiltered: [Currency] = []
    private var currencies = [Currency]()
    private var currenciesFiltered = [Currency]()

    init(
        view: CurrencyPickerView,
        wireframe: CurrencyPickerWireframe,
        interactor: CurrencyPickerInteractor,
        context: CurrencyPickerWireframeContext
    ) {
        self.view = view
        self.wireframe = wireframe
        self.interactor = interactor
        self.context = context
    }
}

extension DefaultCurrencyPickerPresenter: CurrencyPickerPresenter {

    func present() {
        loadSelectedNetworksIfNeeded()
        loadSelectedCurrenciesIfNeeded()
        refreshCurrencies()
        refreshData()
    }

    func handle(_ event: CurrencyPickerPresenterEvent) {
        switch event {
        case let .search(searchTerm):
            self.searchTerm = searchTerm
            refreshData()
        case let .selectNetwork(network):
            guard let network = findSelectedNetwork(from: network) else { return }
            selectedNetwork = network
            selectedCurrencies = interactor.myCurrencies(for: selectedNetwork)
            refreshCurrencies()
            refreshData()
        case let .selectCurrency(currency):
            switch context.source {
            case .multiSelectEdit:
                handleCurrencyTappedOnMultiSelect(currency: currency)
            case let .select(onCompletion):
                guard let currency = findSelectedCurrency(from: currency) else { return }
                onCompletion(selectedNetwork, currency)
            }
        case .addCustomCurrency:
            guard let network = networks.first else { return }
            wireframe.navigate(to: .addCustomCurrency(network: network))
        case .viewWillDismiss:
            guard
                case let CurrencyPickerWireframeContext.Source.multiSelectEdit(_, onCompletion) = context.source
            else { return }
            onCompletion(selectedCurrencies)
        case .done:
            wireframe.dismiss()
        case .dismiss:
            wireframe.dismiss()
        }
    }
}

private extension DefaultCurrencyPickerPresenter {
    
    func loadSelectedNetworksIfNeeded() {
        networks = loadNetworks()
        if let selectedNetwork = context.selectedNetwork, networks.contains(selectedNetwork) {
            self.selectedNetwork = selectedNetwork
        } else {
            selectedNetwork = networks[0]
        }
    }
    
    func loadNetworks() -> [Network] {
        switch context.networks {
        case .all: return interactor.supportedNetworks
        case let .subgroup(networks): return networks
        }
    }
    
    func loadSelectedCurrenciesIfNeeded() {
        switch context.source {
        case let .multiSelectEdit(selectedCurrencies, _):
            self.selectedCurrencies = selectedCurrencies
        default:
            selectedCurrencies = interactor.myCurrencies(for: selectedNetwork)
        }
    }
    
    func handleCurrencyTappedOnMultiSelect(currency: CurrencyPickerViewModel.Currency) {
        if let currency = selectedCurrenciesFiltered.findCurrency(matching: currency.id) {
            selectedCurrencies = selectedCurrencies.removingCurrency(id: currency.id())
        } else if let currency = currenciesFiltered.findCurrency(matching: currency.id) {
            selectedCurrencies = selectedCurrencies.addingCurrency(currency)
        }
        refreshData()
    }
    
    func findSelectedNetwork(from network: CurrencyPickerViewModel.Network) -> Network? {
        networks.first { $0.id() == network.networkId }
    }
    
    func findSelectedCurrency(from currency: CurrencyPickerViewModel.Currency) -> Currency? {
        if let currency = selectedCurrenciesFiltered.findCurrency(matching: currency.id) {
            return currency
        } else if let currency = currenciesFiltered.findCurrency(matching: currency.id) {
            return currency
        } else {
            return nil
        }
    }
    
    func refreshCurrencies() {
        currencies = interactor.currencies(
            filteredBy: searchTerm,
            for: selectedNetwork
        )
    }
    
    func refreshData() {
        selectedCurrenciesFiltered = selectedCurrencies.filterBy(
            searchTerm: searchTerm
        )
        currenciesFiltered = interactor.currencies(
            filteredBy: searchTerm,
            for: selectedNetwork
        )
        updateView(with: sectionsToDisplay())
    }
    
    func updateView(with sectionsDisplayed: [CurrencyPickerViewModel.Section]) {
        let viewModel = viewModel(with: sectionsDisplayed)
        view?.update(with: viewModel)
    }
    
    func viewModel(with sectionsDisplayed: [CurrencyPickerViewModel.Section]) -> CurrencyPickerViewModel {
        .init(
            title: Localized("currencyPicker.title.\(context.title.rawValue)"),
            allowMultiSelection: context.source.isMultiSelect,
            showAddCustomCurrency: context.showAddCustomCurrency,
            content: .loaded(sections: sectionsDisplayed)
        )
    }
    
    func sectionsToDisplay() -> [CurrencyPickerViewModel.Section] {
        var sections = [CurrencyPickerViewModel.Section]()
        if networks.count > 1 {
            sections.append(
                .init(
                    name: Localized("currencyPicker.networks.title"),
                    type: .networks,
                    items: viewModelNetworks()
                )
            )
        }
        if !selectedCurrenciesFiltered.isEmpty {
            sections.append(
                .init(
                    name: Localized("currencyPicker.myTokens.title"),
                    type: .tokens,
                    items: myViewModelCurrencies(from: selectedCurrenciesFiltered)
                )
            )
        }
        if !currenciesFiltered.isEmpty {
            sections.append(
                .init(
                    name: Localized("currencyPicker.other.title"),
                    type: .tokens,
                    items: otherViewModelCurrencies(from: currenciesFiltered)
                )
            )
        }
        return sections.addNoResultsIfNeeded
    }
}

private extension DefaultCurrencyPickerPresenter {
    
    func viewModelNetworks() -> [CurrencyPickerViewModel.Item] {
        networks.compactMap {
            .network(
                .init(
                    networkId: $0.id(),
                    iconName: $0.iconName,
                    name: $0.name,
                    isSelected: $0.id() == selectedNetwork.id()
                )
            )
        }
    }
    
    func myViewModelCurrencies(from currencies: [Currency]) -> [CurrencyPickerViewModel.Item] {
        currencies.compactMap { currency in
            let currencyBalance = interactor.balance(for: currency, network: selectedNetwork)
            let fiatBalancce = currency.fiatValue(for: currencyBalance)
            let type: CurrencyPickerViewModel.CurrencyType
            switch context.source {
            case .multiSelectEdit:
                let isSelected = selectedCurrenciesFiltered.contains(
                    where: {
                        $0.symbol == currency.symbol
                    }
                )
                type = .init(
                    isSelected: isSelected,
                    balance: .init(
                        tokens: Formatters.Companion.shared.currency.format(
                            amount: currencyBalance,
                            currency: currency,
                            style: Formatters.StyleCustom(maxLength: 15)
                        ),
                        usdTotal: Formatters.Companion.shared.fiat.format(
                            amount: fiatBalancce.bigDec,
                            style: Formatters.StyleCustom(maxLength: 10),
                            currencyCode: "usd"
                        )
                    )
                )
            case .select:
                type = .init(
                    isSelected: nil,
                    balance: .init(
                        tokens: Formatters.Companion.shared.currency.format(
                            amount: currencyBalance,
                            currency: currency,
                            style: Formatters.StyleCustom(maxLength: 15)
                        ),
                        usdTotal: Formatters.Companion.shared.fiat.format(
                            amount: fiatBalancce.bigDec,
                            style: Formatters.StyleCustom(maxLength: 10),
                            currencyCode: "usd"
                        )
                    )
                )
            }
            let position: CurrencyPickerViewModel.Currency.Position
            if currencies.first == currency && currencies.last == currency {
                position = .onlyOne
            } else if currencies.first == currency {
                position = .first
            } else if currencies.last == currency {
                position = .last
            } else {
                position = .middle
            }
            return .currency(
                .init(
                    id: currency.coinGeckoId ?? "",
                    imageName: currency.iconName,
                    symbol: currency.symbol,
                    name: currency.name,
                    type: type,
                    position: position
                )
            )
        }
    }
    
    func otherViewModelCurrencies(from currencies: [Currency]) -> [CurrencyPickerViewModel.Item] {
        currencies.compactMap { currency in
            let type: CurrencyPickerViewModel.CurrencyType
            switch context.source {
            case .select:
                type = .init(
                    isSelected: nil,
                    balance: nil
                )
            case .multiSelectEdit:
                let isSelected = selectedCurrenciesFiltered.contains{
                    $0.symbol == currency.symbol
                }
                type = .init(
                    isSelected: isSelected,
                    balance: nil
                )
            }
            let position: CurrencyPickerViewModel.Currency.Position
            if currencies.first == currency && currencies.last == currency {
                position = .onlyOne
            } else if currencies.first == currency {
                position = .first
            } else if currencies.last == currency {
                position = .last
            } else {
                position = .middle
            }
            return .currency(
                .init(
                    id: currency.coinGeckoId ?? "",
                    imageName: currency.iconName,
                    symbol: currency.symbol,
                    name: currency.name,
                    type: type,
                    position: position
                )
            )
        }
    }
}

private extension Array where Element == Network {
    
    func hasNetwork(matching name: String) -> Bool {
        findNetwork(matching: name) != nil
    }
    
    func findNetwork(matching name: String) -> Network? {
        filter { $0.name == name }.first
    }
}

private extension Array where Element == Currency {

    func addingCurrency(_ currency: Currency) -> [Currency] {
        var currencies = self
        currencies.append(currency)
        return currencies
    }
    
    func removingCurrency(id: String) -> [Currency] {
        var currencies = self
        currencies.removeAll {
            $0.coinGeckoId == id
        }
        return currencies
    }
    
    func removingCurrencies(matching currencies: [Currency]) -> [Currency] {
        let ids = currencies.compactMap { $0.coinGeckoId }
        return filter { !ids.contains($0.coinGeckoId ?? "") }
    }
    
    func hasCurrency(with symbol: String) -> Bool {
        findCurrency(with: symbol) != nil
    }
    
    func findCurrency(with symbol: String) -> Currency? {
        filter { $0.symbol == symbol }.first
    }
    
    func findCurrency(matching id: String) -> Currency? {
        
        filter { $0.coinGeckoId == id }.first
    }
}

private extension Array where Element == CurrencyPickerViewModel.Section {
    
    var addNoResultsIfNeeded: [CurrencyPickerViewModel.Section] {
        guard isEmpty else { return self }
        return [
            .init(
                name: Localized("currencyPicker.noResults"),
                type: .tokens,
                items: []
            )
        ]
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/currencyPicker/CurrencyPickerPresenter.kt
package com.sonsofcrypto.web3walletcore.modules.currencyPicker

import com.sonsofcrypto.web3lib.formatters.Formater
import com.sonsofcrypto.web3lib.formatters.Formater.Style.Custom
import com.sonsofcrypto.web3lib.types.Currency
import com.sonsofcrypto.web3lib.types.Network
import com.sonsofcrypto.web3lib.utils.BigDec
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3walletcore.extensions.Localized
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerViewModel.Position
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerViewModel.Position.FIRST
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerViewModel.Position.LAST
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerViewModel.Position.MIDDLE
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerViewModel.Position.SINGLE
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerViewModel.Section.Currencies
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerViewModel.Section.FavouriteCurrencies
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerViewModel.Section.Networks
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerWireframeContext.Result
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerWireframeDestination.AddCustomCurrency
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerWireframeDestination.Dismiss

sealed class CurrencyPickerPresenterEvent {
    data class Search(val searchTerm: String): CurrencyPickerPresenterEvent()
    data class SelectNetwork(val idx: Int): CurrencyPickerPresenterEvent()
    data class SelectFavouriteCurrency(val idx: Int): CurrencyPickerPresenterEvent()
    data class SelectCurrency(val idx: Int): CurrencyPickerPresenterEvent()
    object AddCustomCurrency: CurrencyPickerPresenterEvent()
    object WillDismiss: CurrencyPickerPresenterEvent()
    object Dismiss: CurrencyPickerPresenterEvent()
}

interface CurrencyPickerPresenter {
    fun present()
    fun handle(event: CurrencyPickerPresenterEvent)
}

class DefaultCurrencyPickerPresenter(
    private val view: WeakRef<CurrencyPickerView>,
    private val wireframe: CurrencyPickerWireframe,
    private val interactor: CurrencyPickerInteractor,
    private val context: CurrencyPickerWireframeContext,
): CurrencyPickerPresenter {
    private var searchTerm: String = ""
    private var networks: List<Network> = emptyList()
    private var selectedNetwork: Network
    private var favouriteCurrencies: List<Currency> = emptyList()
    private var currencies: List<Currency> = emptyList()
    private var selectedCurrencies = mutableMapOf<String, List<Currency>>()

    init {
        networks = context.networksData.map { it.network }
        selectedNetwork = context.selectedNetwork ?: networks.first()
        context.networksData.forEach {
            selectedCurrencies[it.network.id()] = it.favouriteCurrencies
                ?: interactor.favouriteCurrencies(it.network)
        }
        refreshCurrencies()
    }

    override fun present() { updateView() }

    override fun handle(event: CurrencyPickerPresenterEvent) {
        when (event) {
            is CurrencyPickerPresenterEvent.Search -> {
                searchTerm = event.searchTerm
                refreshCurrencies()
                updateView()
            }
            is CurrencyPickerPresenterEvent.SelectNetwork -> {
                selectedNetwork = networks[event.idx]
                refreshCurrencies()
                updateView()
            }
            is CurrencyPickerPresenterEvent.SelectFavouriteCurrency -> {
                selectCurrency(favouriteCurrencies[event.idx])
            }
            is CurrencyPickerPresenterEvent.SelectCurrency -> {
                selectCurrency(currencies[event.idx])
            }
            is CurrencyPickerPresenterEvent.AddCustomCurrency -> {
                wireframe.navigate(AddCustomCurrency(selectedNetwork))
            }
            is CurrencyPickerPresenterEvent.WillDismiss -> {
                if (!context.isMultiSelect) return
                context.handler(
                    listOf(Result(selectedNetwork, selectedCurrenciesForSelectedNetwork))
                )
            }
            is CurrencyPickerPresenterEvent.Dismiss -> {
                wireframe.navigate(Dismiss)
            }
        }
    }

    private fun selectCurrency(currency: Currency) {
        if (context.isMultiSelect) {
            val currencies = selectedCurrenciesForSelectedNetwork
            if (currencies.contains(currency)) { currencies.remove(currency) }
            else { currencies.add(currency) }
            selectedCurrencies[selectedNetwork.id()] = currencies
            refreshCurrencies()
            updateView()
        } else {
            context.handler(listOf(Result(selectedNetwork, listOf(currency))))
        }
    }

    private val selectedCurrenciesForSelectedNetwork get() =
        selectedCurrencies[selectedNetwork.id()]?.toMutableList() ?: mutableListOf()

    private fun refreshCurrencies() {
        favouriteCurrencies = filteredFavouriteCurrencies()
        currencies = filteredCurrencies()
    }

    private fun networkData(): CurrencyPickerWireframeContext.NetworkData =
        context.networksData.first { it.network.id() == selectedNetwork.id() }

    private fun filteredFavouriteCurrencies(): List<Currency> =
        filterCurrencies(selectedCurrenciesForSelectedNetwork)

    private fun filteredCurrencies(): List<Currency> {
        val currencies = networkData().currencies
            ?: return interactor.currencies(searchTerm, networkData().network)
        return filterCurrencies(currencies)
    }

    private fun filterCurrencies(currencies: List<Currency>): List<Currency> =
        if (searchTerm.isEmpty()) currencies
        else {
            currencies.filter {
                it.name.contains(searchTerm, ignoreCase = true)
                || it.symbol.contains(searchTerm, ignoreCase = true)
            }
        }

    private fun updateView() {
        view.get()?.update(viewModel())
    }

    private fun viewModel() =
        CurrencyPickerViewModel(
            title(),
            context.isMultiSelect,
            context.showAddCustomCurrency,
            sectionsViewModel()
        )

    private fun title(): String =
        if (context.isMultiSelect) { Localized("currencyPicker.title.currencies") }
        else Localized("currencyPicker.title.currency")

    private fun sectionsViewModel(): List<CurrencyPickerViewModel.Section> {
        val list = mutableListOf<CurrencyPickerViewModel.Section>()
        if (networks.count() > 1) list.add(Networks(sectionNetworksViewModel()))
        if (favouriteCurrencies.isNotEmpty())
            list.add(FavouriteCurrencies(favouriteCurrenciesViewModel()))
        if (currencies.isNotEmpty()) { list.add(Currencies(currenciesViewModel())) }
        return list
    }

    private fun sectionNetworksViewModel(): List<CurrencyPickerViewModel.Network> =
        networks.map {
            CurrencyPickerViewModel.Network(
                "token_eth_icon", it.name, it.id() == selectedNetwork.id()
            )
        }

    private fun favouriteCurrenciesViewModel(): List<CurrencyPickerViewModel.Currency> =
        favouriteCurrencies.map {
            CurrencyPickerViewModel.Currency(
                it.id(),
                it.coinGeckoId ?: "currency_placeholder",
                it.symbol,
                it.name,
                currencyPosition(favouriteCurrencies, it),
                if (context.isMultiSelect) true else null,
                Formater.Companion.currency.format(
                    interactor.balance(selectedNetwork, it), it, Custom(15u),
                ),
                Formater.Companion.fiat.format(
                    BigDec.Companion.from(interactor.fiatPrice(selectedNetwork, it)),
                    Custom(10u), "usd",
                )
            )
        }

    private fun currenciesViewModel(): List<CurrencyPickerViewModel.Currency> =
        currencies.map {
            var isSelected: Boolean? = null
            if (context.isMultiSelect) {
                isSelected = selectedCurrenciesForSelectedNetwork.contains(it)
            }
            CurrencyPickerViewModel.Currency(
                it.id(),
                it.coinGeckoId ?: "currency_placeholder",
                it.symbol,
                it.name,
                currencyPosition(currencies, it),
                isSelected,
                null,
                null
            )
        }

    private fun currencyPosition(list: List<Currency>, item: Currency): Position =
        if (list.first() == item && list.last() == item) { SINGLE }
        else if (list.first() == item) { FIRST }
        else if (list.last() == item) { LAST }
        else { MIDDLE }
}